File: netsim/src/ssfnet/proto/cnf/test/cnf_traffic.m
/**
 * \file cnf_traffic.m
 * \author Ting Li
 * \author Rong Rong
 * 
 * Copyright (c) 2011 Florida International University.
 *
 * Permission is hereby granted, free of charge, to any individual or
 * institution obtaining a copy of this software and associated
 * documentation files (the "software"), to use, copy, modify, and
 * distribute without restriction.
 *
 * The software is provided "as is", without warranty of any kind,
 * express or implied, including but not limited to the warranties of
 * merchantability, fitness for a particular purpose and
 * non-infringement.  In no event shall Florida International
 * University be liable for any claim, damages or other liability,
 * whether in an action of contract, tort or otherwise, arising from,
 * out of or in connection with the software or the use or other
 * dealings in the software.
 *
 * This software is developed and maintained by
 *
 *   Modeling and Networking Systems Research Group
 *   School of Computing and Information Sciences
 *   Florida International University
 *   Miami, Florida 33199, USA
 *
 * You can find our research at http://www.primessf.net/.
 */

#ifndef __CNF_TRAFFIC_H__
#define __CNF_TRAFFIC_H__

#include "os/ssfnet.h"
#include "os/config_type.h"
#include "os/config_entity.h"
#include "os/traffic.h"
#include "os/virtual_time.h"
#include "ssf.h"
#include "os/ssfnet_types.h"

namespace prime {
namespace ssfnet {

// Some value that is unique
#define CNF_TRAFFIC 5008

class CNFStartTrafficEvent: public StartTrafficEvent {
public:
        /** The default constructor. */
        CNFStartTrafficEvent();

        /** The constructor. */
        CNFStartTrafficEvent(UID_t trafficTypeUID, uint32_t id, UID_t h_uid, UID_t d_uid, uint32_t d_ip, VirtualTime tm);

        /** The constructor. */
        CNFStartTrafficEvent(TrafficType* trafficType, uint32_t id, UID_t h_uid, UID_t d_uid, uint32_t d_ip, VirtualTime tm);

        /** The ssf clone constructor. */
        CNFStartTrafficEvent(CNFStartTrafficEvent* evt);

        /** The copy constructor. */
        CNFStartTrafficEvent(const CNFStartTrafficEvent& evt);

        /** The destructor. */
        virtual ~CNFStartTrafficEvent();

        virtual Event* clone() {
                return new CNFStartTrafficEvent(*this);
        }

        virtual prime::ssf::ssf_compact* pack();

        virtual void unpack(prime::ssf::ssf_compact* dp);

        /* The content id. */
        void setContentID(int content_id);

        /* Called by the host.*/
        int getContentID();

        /* The dst port. */
        void setDstPort(int dst_port);

        /* Called by the host.*/
        UID_t getDstPort();

private:
		int content_id;
        UID_t dst_port; //dst port may be used to distinguish different packet /query, reply, regular
public:
        /**
         * This method is the factory method for deserializing the event
         * object. A new traffic event object is created.
         */
        static prime::ssf::Event* createCNFStartTrafficEvent(prime::ssf::ssf_compact* dp);

        /* This macros is used by SSF to declare an event class. */
        SSF_DECLARE_EVENT(CNFStartTrafficEvent);
};


class CNFTraffic: public ConfigurableEntity<CNFTraffic, StaticTrafficType, CNF_TRAFFIC> {
	friend class StartTrafficEvent;
public:
	typedef SSFNET_LIST(UID_t) UIDList;
	typedef SSFNET_VECTOR(float) PopularityVec;
 public:
	state_configuration {
		shared configurable UID_t dst_port {
			type=INT;
			default_value="1000";
			doc_string="The destination port for an CNF connection.";
		};
		shared configurable int content_id {
			type=INT;
			default_value="0";
			doc_string="The contenet id request to the controller.";
		};
		shared configurable int number_of_contents {
			type=INT;
			default_value="1";
			doc_string="The total number of contents.";
		};
		shared configurable int number_of_requests {
			type=INT;
			default_value="1";
			doc_string="The total number of requests.";
		};
	};

	// The constructor
	CNFTraffic();

	// The deconstructor
	virtual ~CNFTraffic();

	ConfigType* getProtocolType();

	/*
	 * the traffic manager will call this to find out if this traffic type should be included
	 */
	virtual bool shouldBeIncludedInCommunity(Community* com);

    /*
     * called when the traffic type is initially created
     */
    virtual void init(){}

	/*
	 * Used by the tm get see if the traffic type needs to start events
	 */
	virtual void getNextEvent(StartTrafficEvent*& traffics_to_start, //the tm will send this to the correct host at the correcT time
			UpdateTrafficTypeEvent*& update_vent, //if there needs to be an update event the TM will deliver it to all concerned traffic types on remote TMs
			bool& wrap_up, //if this is true the TM will wrap up the traffic type and remove it from its active list
			VirtualTime& recall_at //when should the traffic type be recalled -- may need to update its current recall time. if its zero don't change it
			);

    /**
     * The method is used to finish the traffic
     */
	virtual void processFinishedEvent(FinishedTrafficEvent* finished_evt){}

	/*
	 * Return the destination port to connect to
	 */
	UID_t getDstPort();

	/*
	 * Return the file size to request from server
	 */
	int getContentID();
	
	/*
	 * Return the popularity of the content by its cid and the total number of the contents.
	 */
	static float calPopularity(int cid, int num_of_contents);
	
	/*
	 * Return the total number of contents.
	 */
	int getNumberOfContents();
	
	/*
	 * Return the total number of requests.
	 */
	int getNumberOfRequests();
	
	/*
	 * Return the popularity list.
	 */
	static PopularityVec Zipf(float theta, int N);

 private:
	uint32_t traffic_id;
	UIDList srcs_to_use;
	static PRIME_OFSTREAM* prob_file;
	prime::rng::Random * ack_rng;
};

}
}

#endif /* __CNF_TRAFFIC_H__ */

File: netsim/src/ssfnet/proto/cnf/test/cnf_traffic.cc
/**
 * \file cnf_traffic.cc
 * \author Ting Li
 * 
 * Copyright (c) 2011 Florida International University.
 *
 * Permission is hereby granted, free of charge, to any individual or
 * institution obtaining a copy of this software and associated
 * documentation files (the "software"), to use, copy, modify, and
 * distribute without restriction.
 *
 * The software is provided "as is", without warranty of any kind,
 * express or implied, including but not limited to the warranties of
 * merchantability, fitness for a particular purpose and
 * non-infringement.  In no event shall Florida International
 * University be liable for any claim, damages or other liability,
 * whether in an action of contract, tort or otherwise, arising from,
 * out of or in connection with the software or the use or other
 * dealings in the software.
 *
 * This software is developed and maintained by
 *
 *   Modeling and Networking Systems Research Group
 *   School of Computing and Information Sciences
 *   Florida International University
 *   Miami, Florida 33199, USA
 *
 * You can find our research at http://www.primessf.net/.
 */
#include "os/logger.h"
#include "os/ssfnet.h"
#include "os/config_type.h"
#include "os/config_entity.h"
#include "os/traffic.h"
#include "proto/cnf/test/cnf_traffic.h"
#include "os/protocol_session.h"
#include "proto/cnf/cnf_session.h"
#include "os/partition.h"

#define REQUEST_INTERVAL 0.1

namespace prime {
namespace ssfnet {

LOGGING_COMPONENT(CNFTraffic);

PRIME_OFSTREAM* CNFTraffic::prob_file=0;


CNFStartTrafficEvent::CNFStartTrafficEvent() :
		StartTrafficEvent(SSFNetEvent::SSFNET_TRAFFIC_CNF_START_EVT), content_id(0), dst_port(0) {
}
CNFStartTrafficEvent::CNFStartTrafficEvent(UID_t traffic_type_uid, uint32_t id,
		UID_t h_uid, UID_t d_uid, uint32_t d_ip, VirtualTime st) :
		StartTrafficEvent(SSFNetEvent::SSFNET_TRAFFIC_CNF_START_EVT,
				traffic_type_uid, id, h_uid, d_uid, d_ip, st), content_id(0), dst_port(0) {
}
CNFStartTrafficEvent::CNFStartTrafficEvent(TrafficType* traffic_type,
		uint32_t id, UID_t h_uid, UID_t d_uid, uint32_t d_ip, VirtualTime st) :
		StartTrafficEvent(SSFNetEvent::SSFNET_TRAFFIC_CNF_START_EVT,
				traffic_type, id, h_uid, d_uid, d_ip, st), content_id(0), dst_port(0) {
}
CNFStartTrafficEvent::CNFStartTrafficEvent(CNFStartTrafficEvent* evt) :
		StartTrafficEvent(evt), content_id(evt->content_id), dst_port(
				evt->dst_port) {
}
CNFStartTrafficEvent::CNFStartTrafficEvent(const CNFStartTrafficEvent& evt) :
		StartTrafficEvent(evt), content_id(evt.content_id), dst_port(
				evt.dst_port) {
}
CNFStartTrafficEvent::~CNFStartTrafficEvent() {

}
prime::ssf::ssf_compact* CNFStartTrafficEvent::pack() {
	prime::ssf::ssf_compact* dp = StartTrafficEvent::pack();
	dp->add_unsigned_long_long(content_id);
	dp->add_unsigned_long_long(dst_port);
	return dp;
}

void CNFStartTrafficEvent::unpack(prime::ssf::ssf_compact* dp) {
	StartTrafficEvent::unpack(dp);
	dp->get_int(&content_id, 1);
	dp->get_unsigned_long_long(&dst_port, 1);
}

void CNFStartTrafficEvent::setContentID(int c_id) {
	content_id = c_id;
}

int CNFStartTrafficEvent::getContentID() {
	return content_id;
}

void CNFStartTrafficEvent::setDstPort(int p) {
	dst_port = p;
}

UID_t CNFStartTrafficEvent::getDstPort() {
	return dst_port;
}

prime::ssf::Event* CNFStartTrafficEvent::createCNFStartTrafficEvent(
		prime::ssf::ssf_compact* dp) {
	CNFStartTrafficEvent* t_evt = new CNFStartTrafficEvent();
	t_evt->unpack(dp);
	return t_evt;
}

ConfigType* CNFTraffic::getProtocolType() {
	return CNFApplication::getClassConfigType();
}

CNFTraffic::CNFTraffic() :
		traffic_id(0) {
	int seed = (int) rand();
	ack_rng=new prime::rng::Random(prime::rng::Random::RNG_DASSF_LEHMER, seed);
	ack_rng->uniform();

	if(!prob_file) {
		prob_file = new PRIME_OFSTREAM();
		prob_file->open("prob");
	}
}

CNFTraffic::~CNFTraffic(){}

bool CNFTraffic::shouldBeIncludedInCommunity(Community* com) {
	return StaticTrafficType::shouldBeIncludedInCommunity(com);
}

void CNFTraffic::getNextEvent(StartTrafficEvent*& traffics_to_start, //the tm will send this to the correct host at the correcT time
		UpdateTrafficTypeEvent*& update_evt, //if there needs to be an update event the TM will deliver it to all concerned traffic types on remote TMs
		bool& wrap_up, //if this is true the TM will wrap up the traffic type and remove it from its active list
		VirtualTime& recall_at //when should the traffic type be recalled -- may need to update its current recall time. if its zero don't change it
		) {
	traffics_to_start = new CNFStartTrafficEvent();
	if (traffic_id == 0) {
		if (traffics_to_start) {
			//LOG_DEBUG("traffic_id="<<traffic_id<<endl)
			UIDVec& src_rids = shared.srcs.read().getCompiledRI()->getUIDVec();
			LOG_DEBUG("The size of the srcs vector ="<<src_rids.size()<<endl);
			int num_of_contents = getNumberOfContents();
			int num_of_requests = getNumberOfRequests()*3;
			int content_id = getContentID();
			UIDVec src_uids;
			BaseEntity* anchor = getParent()->getParent();
			for (UIDVec::iterator it = src_rids.begin(); it < src_rids.end(); it++) {
				src_uids.push_back(*it+(anchor->getUID()-anchor->getSize()));
			}
			LOG_DEBUG("The size of the srcs uids ="<<src_uids.size()<<endl);
			//calculate the popularity of the content
			//select a number of srcs_to_use from src_rids according to the popularity
			//push srcs to a list, every time pop one src to send request
			//while the list is empty, wrap up the traffic.
			//HACK, for each content, calculate its popularity
			float popularity = this->calPopularity(content_id,num_of_contents);
			content_id += 180*(rand()%5);//XXX HACK
			int requests_to_send = (int)num_of_requests*popularity;
			if(requests_to_send==0){
				requests_to_send = 1;
			}
			LOG_DEBUG("The number of requests to send ="<<requests_to_send<<endl);
			for(int i=0; i<requests_to_send; i++){
				int index = rand() % src_uids.size();
				srcs_to_use.push_back(src_uids[index]);
			}
			LOG_DEBUG("The size of the srcs to use ="<<srcs_to_use.size()<<endl);
			UID_t src_uid = srcs_to_use.front();
			srcs_to_use.pop_front();
			LOG_DEBUG("CNF_TRAFFIC, src uid is: "<<src_uid<<endl)
			traffics_to_start->setStartTime(VirtualTime(shared.start_time.read(), VirtualTime::SECOND));
			traffics_to_start->setHostUID(src_uid);
			traffics_to_start->setDstUID(0);
			traffics_to_start->setTrafficId(traffic_id++);
			traffics_to_start->setTrafficType(this);
			traffics_to_start->setTrafficTypeUID(getUID());

			//set the additional parameters in CNF start traffic event: content id, dst port
			SSFNET_DYNAMIC_CAST(CNFStartTrafficEvent*,traffics_to_start)->setContentID(
					content_id);
			SSFNET_DYNAMIC_CAST(CNFStartTrafficEvent*,traffics_to_start)->setDstPort(
					getDstPort());

			if(prob_file)
				(*prob_file)<<"CNFPROBsrc=="<<src_uid<<"==content=="<<content_id<<endl;

			if (srcs_to_use.empty()) {
				wrap_up = true;
			} else {
				if(shared.interval_exponential.read()){
					recall_at = VirtualTime(shared.start_time.read(), VirtualTime::SECOND)
							+ VirtualTime(ack_rng->exponential(1.0/shared.interval.read()), VirtualTime::SECOND);
				}else{
					recall_at = VirtualTime(shared.start_time.read(), VirtualTime::SECOND)
							+ VirtualTime(shared.interval.read(), VirtualTime::SECOND);
				}
			}
		}
	} else {
		//int content_id = getContentID();
		int content_id = getContentID()+180*(rand()%5);//XXX HACK
		UID_t src_uid = srcs_to_use.front();
		srcs_to_use.pop_front();
		traffics_to_start->setStartTime(VirtualTime(prime::ssf::now()));
		traffics_to_start->setHostUID(src_uid);
		traffics_to_start->setDstUID(0);
		traffics_to_start->setTrafficId(traffic_id++);
		traffics_to_start->setTrafficType(this);
		traffics_to_start->setTrafficTypeUID(getUID());

		//set the additional parameters in CNF start traffic event: content id, dst port
		SSFNET_DYNAMIC_CAST(CNFStartTrafficEvent*,traffics_to_start)->setContentID(
				content_id);
		SSFNET_DYNAMIC_CAST(CNFStartTrafficEvent*,traffics_to_start)->setDstPort(
				getDstPort());
		if(prob_file)
			(*prob_file)<<"CNFPROBsrc=="<<src_uid<<"==content=="<<getContentID()<<endl;

		if (srcs_to_use.empty()) {
			wrap_up = true;
		} else {
			if(shared.interval_exponential.read()){
				recall_at = VirtualTime(ack_rng->exponential(1.0/shared.interval.read()), VirtualTime::SECOND);
			}else{
				recall_at = VirtualTime(shared.interval.read(), VirtualTime::SECOND);
			}
		}
	}
}

int CNFTraffic::getContentID() {
	return shared.content_id.read();
}

UID_t CNFTraffic::getDstPort() {
	return shared.dst_port.read();
}

float CNFTraffic::calPopularity(int cid, int num_of_contents){
	PopularityVec p = Zipf(0.5, num_of_contents);
	return p[cid-1];
}

int CNFTraffic::getNumberOfContents() {
	return shared.number_of_contents.read();
}

int CNFTraffic::getNumberOfRequests() {
	return shared.number_of_requests.read();
}

CNFTraffic::PopularityVec CNFTraffic::Zipf(float theta, int N){
	/*
	 * zipfian - p(i) = c / i ^^ (1 - theta)
	 * At theta = 1, uniform *
	 * at theta = 0, pure zipfian
	 */
    float sum=0.0;
    float c=0.0;
    float expo;
    int i;
    PopularityVec zdist(N);
    expo = 1 - theta;

    for (i = 1; i <= N; i++) {
        sum += 1.0 /(float) pow((double) i, (double) (expo));
    }
    c = 1.0 / sum;

    for (i = 0; i < N; i++) {
         zdist[i] = c / (float) pow((double) (i + 1), (double) (expo));
    }

    return zdist;
}

// ssf requires this macro to register an event class.
SSF_REGISTER_EVENT(CNFStartTrafficEvent,
		CNFStartTrafficEvent::createCNFStartTrafficEvent);

} // namespace ssfnet
} // namespace prime

